### DISCUSSION TITLE
Use cases: Materialize + Hasura

### DISCUSSION THREAD

**Discussion by rrjanbiah** [MaterializeInc/materialize#5764] [Q&A]
Would like to know more about the use cases of 'Materialize + Hasura' from the community. I came through https://github.com/MaterializeInc/materialize/issues/3496 (More specifically, https://github.com/hasura/graphql-engine/issues/4307 for the context)

Basically, this is for the continuation of discussion in https://github.com/MaterializeInc/materialize/issues/4150#issuecomment-778646314 

@ajbouh Wrote:

> I'm not sure it's correct to say someone looking to use Materialized with Hasura is only getting into edge cases or distributed systems. Nor that someone might be starting with a choice of database and then using Hasura because of its compatibility with that database.
> 
> I personally started with Hasura because I wanted to have an edge-accessible backend without writing a lot of manual boilerplate. My choice was either AWS AppSync or Hasura. I'm only using postgres because Hasura uses it.
> 
> As for Materialized, I want to have live updated values visible to my users in-browser. Hasura's built-in subscription implementation is fine for simple queries but not complex views.
> 
> My driving interest in Materialized is to have efficient subscription support for complex (and incrementally computable) views.

Would you mind sharing more about your use cases and challenges?

Will https://github.com/hasura/graphql-engine/issues/2317 address your pain points? (It is supposed to be the fix for https://github.com/hasura/graphql-engine/issues/2907#issuecomment-539477677)

In case, if you have already checked below, are they relevant here? 

1. https://github.com/supabase/realtime
2. https://github.com/pubkey/rxdb
3. https://github.com/lawrencejones/pgsink
4. Different ecosystem, but somewhat relevant (?) here https://github.com/zombodb/zombodb & https://github.com/deepstreamIO/deepstream.io

**Comment by ajbouh** [comment]
Yes, happy to share.

I have two main streaming use cases:

**Statistics**

I have a set of complex queries that compute various statistics from a set of interrelated tables. Some of these statistics are simple aggregations (e.g. mean, max, p99). Others require doing multiple joins to compute the time between two events (e.g. avg time between user sign up and completion of a specific task).

Today I achieve this by:
- writing a complex postgresql view
- querying this view via Hasura during SSR
- using Hasura subscriptions to watch it from the browser for updates

**Recent event stream**

I have a number of UI views that show live updating tables of recent events. For example a log of recent API-initiated actions.

Today I achieve this by:
- querying relevant table via Hasura during SSR
- subscribing to events from the browser for updates, filtering to ones newer than those returned during SSR

**Scaling concerns**
My concerns with the approaches I've taken above are primarily:
- user-visible performance for users with many rows to be aggregated
- the unit economics of scaling up the number of users

**Reply by rrjanbiah** [reply]
@ajbouh Thank you so much for taking time to explain. Very helpful to understand your use case. BTW, did you consider / try https://github.com/timescale/timescaledb?
